File: OptimOption.m
% This class specifies the options for the control parameters of the algorithm.
% Written by Ali Ahrari (aliahrari1983@gmail.com)
% last updated by Ali Ahrari on 16 Jan 2022 

classdef OptimOption < handle
    properties
        archiving; % Object archive includes options for updating the archive
        coreSearch; % options for the core search algorithm
        niching; % options for the niching mechanism
        stopCr; %  stopping criteria for termination of a subpopulation
        dyna; % dynamic options
    end
    methods
        function opt=OptimOption(problem,coreSearchName)
            opt.archiving=Archiving(problem); % Object archive includes options for updating the archive
            if strcmp(coreSearchName,'CMSA')
                opt.coreSearch=CoreSearchCMSA(problem);
            elseif strcmp(coreSearchName,'CMA')
                opt.coreSearch=CoreSearchCMA(problem);
            else
                disp('Option for core search is invalid')
                abort
            end
            opt.niching=Niching(problem); % options for the niching mechanism
            opt.stopCr=StopCriteria(problem); % other stopping criteria for termination of a subpopulation
            opt.dyna=DynamicOption();
        end
    end
end
    
    

File: optim_option.py
"""Dynamic option and optimization option classes."""
from __future__ import annotations
from dataclasses import dataclass
from core_search import CoreSearchCMSA
from data_structures import Niching, Archiving, StopCriteria
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from optim_problem import OptimProblem


@dataclass
class DynamicOption:
    """Options peculiar to dynamic problems."""
    benchSolSizeCoeff: float = 1.0
    recStepSizeCoeff: float = 0.5
    tolChangeF: float = 1e-5
    predictMethod: str = "AMLP"
    chCheckFr: int = 10
    maxPL: int = 4


class OptimOption:
    """Options for the control parameters of the algorithm."""
    
    def __init__(self, problem: OptimProblem, coreSearchName: str = 'CMSA'):
        self.archiving = Archiving(problem)
        if coreSearchName == 'CMSA':
            self.coreSearch = CoreSearchCMSA(problem)
        else:
            raise ValueError(f'Option for core search is invalid: {coreSearchName}')
        self.niching = Niching(problem)
        self.stopCr = StopCriteria(problem)
        self.dyna = DynamicOption()
